Ignore kernel-assigned LL addrs when selecting "bip6"

robmry opened the pull request:
**- What I did**

Noticed that, sometimes (under conditions I've not managed to repro reliably) when enabling IPv6 on the default bridge without specifying a `fixed-cidr-v6`, `docker0` wasn't assigned an address from the daemon's implicit ULA subnet - it only got an IPv4 address and the kernel-assigned link local address.

Commit facb2323 (https://github.com/moby/moby/pull/48319 - not shipped) aligned the way the default bridge's IPv6 subnet and gateway addresses are selected with IPv4.

Part of that involved looking at addresses already on the bridge, along with daemon config options. But, for IPv6, the kernel will assign a link-local address to the bridge - and that was confusing things.

**- How I did it**

_**Run tests that change docker0 in their own netns**_

These tests create iptables rules for different addresses on `docker0` but, unlike tests that do that for user-defined bridges, those rules aren't removed when the test deletes the network, because the default bridge network can't be deleted.

Because of this, adding tests for LL addresses in the third commit in this PR caused `TestAccessPublishedPortFromHost` to fail - it uses a custom bridge and wasn't affected by the daemon change in this PR, but NAT rules left behind by the new tests broke it.

So, use (abuse) the `L3Segment` code to run the tests in their own network namespace.

A minor drawback is that `TestDaemonDefaultBridgeWithFixedCidrButNoBip` can no longer run in rootless mode, because the `L3Segment` can't be created in rootlesskit's netns. But the test isn't looking at anything specific to rootless mode.

_**Remove test env var DOCKER_TEST_CREATE_DEFAULT_BRIDGE**_

Env var `DOCKER_TEST_CREATE_DEFAULT_BRIDGE` could be set to override the name of the default bridge - without the bridge being user-managed (unlike the `--bridge` daemon option).

It was needed by tests looking at docker0 behaviour, using their own instance of the daemon, without breaking the
docker0 instance belonging to CI's daemon. Now, those tests run in their own netns using the name docker0.

So, remove the unused env var.

_**Ignore kernel-assigned LL addrs when selecting "bip6"**_

Make sure the kernel-assigned LL address is ignored when selecting "bip6" (when bip6 is not explicitly specified).

This is made slightly complicated because we allow `fixed-cidr-v6` to be a link-local subnet (either the standard "fe80::/64", or any other non-overlapping LL subnet in "fe80::/10").

Following this change, if fixed-cidr-v6 is (or is included by) "fe80::/64", the bridge's kernel-assigned LL address may be used as the network's gateway address - even though it may also get an IPAM-assigned LL address.

**- How to verify it**

Updated the integration tests to make sure the bridge always has a kernel-assigned LL address - and added test cases for link-local subnets.

Without the fix, two of the updated tests fail ...

```
Running /go/src/github.com/docker/docker/integration/daemon (arm64.integration.daemon) flags=-test.v -test.timeout=10m -test.run TestDaemonDefaultBridgeIPAM
INFO: Testing against a local daemon
=== RUN   TestDaemonDefaultBridgeIPAM_Docker0
=== RUN   TestDaemonDefaultBridgeIPAM_Docker0/no_config
    daemon_linux_test.go:421: assertion failed:
        --- insp.IPAM.Config
        +++ expIPAMConfig
          []network.IPAMConfig{
          	{
        - 		Subnet:     "fe80::/64",
        + 		Subnet:     "192.168.176.0/24",
          		IPRange:    "",
        - 		Gateway:    "fe80::5448:e2ff:fe6a:4500",
        + 		Gateway:    "192.168.176.1",
          		AuxAddress: nil,
          	},
          	{
        - 		Subnet:     "192.168.176.0/24",
        + 		Subnet:     "fdd1:8161:2d2c::/64",
          		IPRange:    "",
        - 		Gateway:    "192.168.176.1",
        + 		Gateway:    "fdd1:8161:2d2c::1/64",
          		AuxAddress: nil,
          	},
          }

=== RUN   TestDaemonDefaultBridgeIPAM_Docker0/fixed-cidr_only
=== RUN   TestDaemonDefaultBridgeIPAM_Docker0/bip_only
=== RUN   TestDaemonDefaultBridgeIPAM_Docker0/existing_bridge_address_only
=== RUN   TestDaemonDefaultBridgeIPAM_Docker0/fixed-cidr_within_old_bridge_subnet
=== RUN   TestDaemonDefaultBridgeIPAM_Docker0/link-local_fixed-cidr-v6
=== RUN   TestDaemonDefaultBridgeIPAM_Docker0/nonstandard_link-local_fixed-cidr-v6
=== RUN   TestDaemonDefaultBridgeIPAM_Docker0/fixed-cidr_within_old_bridge_subnet_with_new_bip
=== RUN   TestDaemonDefaultBridgeIPAM_Docker0/old_bridge_subnet_within_fixed-cidr
=== RUN   TestDaemonDefaultBridgeIPAM_Docker0/old_bridge_subnet_outside_fixed-cidr
=== RUN   TestDaemonDefaultBridgeIPAM_Docker0/old_bridge_subnet_outside_fixed-cidr_with_bip
--- FAIL: TestDaemonDefaultBridgeIPAM_Docker0 (6.55s)
    --- FAIL: TestDaemonDefaultBridgeIPAM_Docker0/no_config (0.60s)
    --- PASS: TestDaemonDefaultBridgeIPAM_Docker0/fixed-cidr_only (0.59s)
    --- PASS: TestDaemonDefaultBridgeIPAM_Docker0/bip_only (0.59s)
    --- PASS: TestDaemonDefaultBridgeIPAM_Docker0/existing_bridge_address_only (0.59s)
    --- PASS: TestDaemonDefaultBridgeIPAM_Docker0/fixed-cidr_within_old_bridge_subnet (0.60s)
    --- PASS: TestDaemonDefaultBridgeIPAM_Docker0/link-local_fixed-cidr-v6 (0.60s)
    --- PASS: TestDaemonDefaultBridgeIPAM_Docker0/nonstandard_link-local_fixed-cidr-v6 (0.60s)
    --- PASS: TestDaemonDefaultBridgeIPAM_Docker0/fixed-cidr_within_old_bridge_subnet_with_new_bip (0.60s)
    --- PASS: TestDaemonDefaultBridgeIPAM_Docker0/old_bridge_subnet_within_fixed-cidr (0.59s)
    --- PASS: TestDaemonDefaultBridgeIPAM_Docker0/old_bridge_subnet_outside_fixed-cidr (0.60s)
    --- PASS: TestDaemonDefaultBridgeIPAM_Docker0/old_bridge_subnet_outside_fixed-cidr_with_bip (0.60s)
=== RUN   TestDaemonDefaultBridgeIPAM_UserBr
=== RUN   TestDaemonDefaultBridgeIPAM_UserBr/bridge_only
=== RUN   TestDaemonDefaultBridgeIPAM_UserBr/fixed-cidr_only
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.378304680Z" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -C DOCKER-USER -j RETURN]"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.378768055Z" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -C FORWARD -j DOCKER-USER]"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.379219805Z" level=debug msg="/usr/sbin/ip6tables, [--wait -D FORWARD -j DOCKER-USER]"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.379704430Z" level=debug msg="/usr/sbin/ip6tables, [--wait -I FORWARD -j DOCKER-USER]"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.380792555Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -n -L DOCKER-USER]"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.381250055Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -C DOCKER-USER -j RETURN]"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.381705305Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -C FORWARD -j DOCKER-USER]"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.382163430Z" level=debug msg="/usr/sbin/iptables, [--wait -D FORWARD -j DOCKER-USER]"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.382651096Z" level=debug msg="/usr/sbin/iptables, [--wait -I FORWARD -j DOCKER-USER]"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.383120971Z" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -n -L DOCKER-USER]"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.383612430Z" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -C DOCKER-USER -j RETURN]"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.384440888Z" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -C FORWARD -j DOCKER-USER]"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.384975930Z" level=debug msg="/usr/sbin/ip6tables, [--wait -D FORWARD -j DOCKER-USER]"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.385480846Z" level=debug msg="/usr/sbin/ip6tables, [--wait -I FORWARD -j DOCKER-USER]"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.386415763Z" level=debug msg="daemon configured with a 15 seconds minimum shutdown timeout"
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.386447055Z" level=debug msg="start clean shutdown of all containers with a 15 seconds timeout..."
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.386827471Z" level=debug msg="Unix socket /tmp/dxr/daf4351ff91cf/libnetwork/940813862d38.sock was closed. The external key listener will stop."
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.387220346Z" level=debug msg="Cleaning up old mountid : start."
    daemon.go:318: [daf4351ff91cf] time="2024-12-03T17:58:32.387999721Z" level=debug msg="Cleaning up old mountid : done."
    daemon.go:318: [daf4351ff91cf] failed to start daemon: Error initializing network controller: fixed-cidr-v6=fdd1:8161:2d2c::/64 is outside any subnet implied by addresses on the user-managed default bridge
    daemon_linux_test.go:409: [daf4351ff91cf] failed to start daemon with arguments [--data-root /go/src/github.com/docker/docker/bundles/test-integration/TestDaemonDefaultBridgeIPAM_UserBr/fixed-cidr_only/daf4351ff91cf/root --exec-root /tmp/dxr/daf4351ff91cf --pidfile /go/src/github.com/docker/docker/bundles/test-integration/TestDaemonDefaultBridgeIPAM_UserBr/fixed-cidr_only/daf4351ff91cf/docker.pid --userland-proxy=true --containerd-namespace daf4351ff91cf --containerd-plugins-namespace daf4351ff91cfp --containerd /var/run/docker/containerd/containerd.sock --host unix:///tmp/docker-integration/daf4351ff91cf.sock --debug --storage-driver vfs --fixed-cidr 192.168.176.0/24 --fixed-cidr-v6 fdd1:8161:2d2c::/64 --ipv6 --bridge br-dbi] : [daf4351ff91cf] daemon exited during startup: exit status 1
    daemon_linux_test.go:399: [daf4351ff91cf] daemon is not started
=== RUN   TestDaemonDefaultBridgeIPAM_UserBr/fcidr_in_bridge_subnet_and_bridge_ip_in_fcidr
=== RUN   TestDaemonDefaultBridgeIPAM_UserBr/fcidr_in_bridge_subnet_and_bridge_ip_not_in_fcidr
=== RUN   TestDaemonDefaultBridgeIPAM_UserBr/link-local_fixed-cidr-v6
=== RUN   TestDaemonDefaultBridgeIPAM_UserBr/nonstandard_link-local_fixed-cidr-v6
=== RUN   TestDaemonDefaultBridgeIPAM_UserBr/fixed-cidr_bigger_than_bridge_subnet
=== RUN   TestDaemonDefaultBridgeIPAM_UserBr/no_bridge_ip_within_fixed-cidr
=== RUN   TestDaemonDefaultBridgeIPAM_UserBr/fixed-cidr_contains_bridge_subnet
=== RUN   TestDaemonDefaultBridgeIPAM_UserBr/fixed-cidr-v6_bigger_than_bridge_subnet
    daemon_linux_test.go:399: [d77287f8daafc] daemon is not started
=== RUN   TestDaemonDefaultBridgeIPAM_UserBr/no_bridge_ip_within_fixed-cidr-v6
    daemon_linux_test.go:399: [dff5af1a1a58e] daemon is not started
=== RUN   TestDaemonDefaultBridgeIPAM_UserBr/fixed-cidr-v6_contains_bridge_subnet
    daemon_linux_test.go:399: [d47d53e1bbe3d] daemon is not started
--- FAIL: TestDaemonDefaultBridgeIPAM_UserBr (7.09s)
    --- PASS: TestDaemonDefaultBridgeIPAM_UserBr/bridge_only (0.60s)
    --- FAIL: TestDaemonDefaultBridgeIPAM_UserBr/fixed-cidr_only (0.56s)
    --- PASS: TestDaemonDefaultBridgeIPAM_UserBr/fcidr_in_bridge_subnet_and_bridge_ip_in_fcidr (0.61s)
    --- PASS: TestDaemonDefaultBridgeIPAM_UserBr/fcidr_in_bridge_subnet_and_bridge_ip_not_in_fcidr (0.60s)
    --- PASS: TestDaemonDefaultBridgeIPAM_UserBr/link-local_fixed-cidr-v6 (0.59s)
    --- PASS: TestDaemonDefaultBridgeIPAM_UserBr/nonstandard_link-local_fixed-cidr-v6 (0.60s)
    --- PASS: TestDaemonDefaultBridgeIPAM_UserBr/fixed-cidr_bigger_than_bridge_subnet (0.59s)
    --- PASS: TestDaemonDefaultBridgeIPAM_UserBr/no_bridge_ip_within_fixed-cidr (0.62s)
    --- PASS: TestDaemonDefaultBridgeIPAM_UserBr/fixed-cidr_contains_bridge_subnet (0.59s)
    --- PASS: TestDaemonDefaultBridgeIPAM_UserBr/fixed-cidr-v6_bigger_than_bridge_subnet (0.58s)
    --- PASS: TestDaemonDefaultBridgeIPAM_UserBr/no_bridge_ip_within_fixed-cidr-v6 (0.58s)
    --- PASS: TestDaemonDefaultBridgeIPAM_UserBr/fixed-cidr-v6_contains_bridge_subnet (0.58s)
FAIL
```

**- Description for the changelog**
```markdown changelog
n/a
```

robmry committed:
Run tests that change docker0 in their own netns

These tests create iptables rules for different addresses on
docker0 but, unlike tests that do that for user-defined bridges,
those rules aren't removed when the test deletes the network,
because the default bridge network can't be deleted.

So, use (abuse) the L3Segment code to run the tests in their
own network namespace.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Remove test env var DOCKER_TEST_CREATE_DEFAULT_BRIDGE

Env var DOCKER_TEST_CREATE_DEFAULT_BRIDGE could be set to
override the name of the default bridge - without the bridge
being user-managed (unlike the '--bridge' daemon option).

It was needed by tests looking at docker0 behaviour, using
their own instance of the daemon, without breaking the
docker0 instance belonging to CI's daemon.

Now, those tests run in their own netns using the name docker0.

So, remove the unused env var.

Signed-off-by: Rob Murray <rob.murray@docker.com>

akerouanton reviewed (commented):
There's an extra debug file that need to go away 😉 Otherwise, LGTM.

robmry committed:
Ignore kernel-assigned LL addrs when selecting "bip6"

Commit facb2323 aligned the way the default bridge's IPv6 subnet
and gateway addresses are selected with IPv4.

Part of that involved looking at addresses already on the bridge,
along with daemon config options. But, for IPv6, the kernel will
assign a link-local address to the bridge.

Make sure that address is ignored when selecting "bip6" when it's
not explicitly specified.

This is made slightly complicated because we allow fixed-cidr-v6
to be a link-local subnet (either the standard "fe80::/64", or
any other non-overlapping LL subnet in "fe80::/10").

Following this change, if fixed-cidr-v6 is (or is included by)
"fe80::/64", the bridge's kernel-assigned LL address may be used
as the network's gateway address - even though it may also get an
IPAM-assigned LL address.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented:
Unrelated test failure ...

```
=== FAIL: github.com/docker/docker/integration/container TestWaitNonBlocked/wait-nonblocking-exit-0 (60.34s)
    wait_test.go:44: assertion failed: error is not nil: error during connect: Post "http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.48/containers/create": EOF
```

thaJeztah commented on integration/daemon/daemon_linux_test.go:
```diff
@@ -3,43 +3,45 @@ package daemon // import "github.com/docker/docker/integration/daemon"
+	host.Do(t, func() {
+		// Run without OTel because there's no routing from this netns for it - which
+		// means the daemon doesn't shut down cleanly, causing the test to fail.
+		d := daemon.New(t, daemon.WithEnvVars("OTEL_EXPORTER_OTLP_ENDPOINT="))
```
Would this be something we could control through the `ctx := testutil.StartSpan(baseContext, t)` utility? It looks like that checks if Otel is configured in the environment; perhaps that could also take an argument to run `t.SetEnv()` for the test (and restore it after). Not  sure if `t.SetEnv` would be problematic if we want to run things in parallel though.


(definitely in favour of having an option for this, so that we don't have to know about specific env-vars everywhere)

thaJeztah commented on daemon/daemon_unix.go:
```diff
@@ -1190,6 +1191,15 @@ func selectBIP(
+	// link-local addresses, unless fixed-cidr-v6 has the standard link-local
+	// prefix. (If fixed-cidr-v6 is the standard LL prefix, the kernel-assigned
+	// address is likely to be used instead of an IPAM assigned address.)
+	if family == netlink.FAMILY_V6 && (fCidrIP == nil || !isStandardLL(fCidrIP)) {
```
Nit (not need to fix); looks like the `fCidrIP == nil` is redundant here, because  `isStandardLL` already returns false if it's `nil`

thaJeztah reviewed (approved):
LGTM
